---
title: 'Error reference'
metaTitle: 'Error reference (Reference)'
metaDescription: 'Prisma Client errors'
---

<TopBlock></TopBlock>

## <inlinecode>PrismaClientKnownRequestError</inlinecode>

As soon as the validation was successful, Prisma Client sends the query request to the query engine. The query engine might now return an error related to the request. This could e.g. be a unique constraint violation while creating a record. All `PrismaClientKnownRequestError` errors have a `code` property and optionally a meta property including relevant meta information to the error.

### Error codes

#### Query Engine

| **Error** | **Example message**                                                                                                                                           | **Description** |
| :-------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------ | :-------------- |
| `P2000`   | "The provided value for the column is too long for the column's type. Column: \${column_name}"<br />                                                          |
| `P2001`   | "The record searched for in the where condition (`${model_name}.${argument_name} = ${argument_value}`) does not exist"<br />                                  |
| `P2003`   | "Foreign key constraint failed on the field: `${field_name}`"<br />                                                                                           |
| `P2005`   | "The value `${field_value}` stored in the database for the field `${field_name}` is invalid for the field's type"<br />                                       |
| `P2006`   | "The provided value `${field_value}` for `${model_name}` field `${field_name}` is not valid"<br />                                                            |
| `P2008`   | "Failed to parse the query `${query_parsing_error}` at `${query_position}`"<br />                                                                             |
| `P2009`   | "Failed to validate the query `${query_validation_error}` at `${query_position}`"<br />                                                                       |
| `P2013`   | "Missing the required argument `${argument_name}` for field `${field_name}` on `${object_name}`."<br />                                                       |
| `P2014`   | "The change you are trying to make would violate the required relation '\${relation_name}' between the `${model_a_name}` and `${model_b_name}` models."<br /> |
| `P2017`   | "The records for relation `${relation_name}` between the `${parent_name}` and `${child_name}` models are not connected."<br />                                |
| `P2018`   | "The required connected records were not found. \${details}"<br />                                                                                            |
| `P2021`   | "The table `${table}` does not exist in the current database."<br />                                                                                          |
| `P2022`   | "The column `${column}` does not exist in the current database."<br />                                                                                        |

#### Migration Engine

| **Error** | **Example message**                                                                                                                                                              | **Description** |
| :-------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :-------------- |
| `P3001`   | "Migration possible with destructive changes and possible data loss: \${migration_engine_destructive_details}"<br />                                                             | `P3002`         | "The attempted migration was rolled back: \${database_error}"<br /> |
| `P3003`   | "The format of migrations changed, the saved migrations are no longer valid. To solve this problem, please follow the steps at: https://pris.ly/d/migrate#troubleshooting"<br /> |
| `P3004`   | "The `${database_name}` database is a system database, it should not be altered with prisma migrate. Please connect to another database."<br />                                  |

#### Common

| **Error** | **Example message**                                                                                                                                                                                                                                                                                                                                                                                                                                                     | **Description** |
| :-------- | :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :-------------- |
| `P1000`   | "Authentication failed against database server at `${database_host}`, the provided database credentials for `${database_user}` are not valid. Please make sure to provide valid database credentials for the database server at `${database_host}`."<br />                                                                                                                                                                                                              | `P1001`         | "Can't reach database server at `${database_host}`:`${database_port}` Please make sure your database server is running at `${database_host}`:`${database_port}`."<br /> |
| `P1002`   | "The database server at `${database_host}`:`${database_port}` was reached but timed out. Please try again. Please make sure your database server is running at `${database_host}`:`${database_port}`. "<br />"Database `${database_name}.${database_schema_name}` does not exist on the database server at `${database_host}:${database_port}`."<br />"Database `${database_name}` does not exist on the database server at `${database_host}:${database_port}`."<br /> |
| `P1009`   | "Database `${database_name}` already exists on the database server at `${database_host}:${database_port}`"<br />                                                                                                                                                                                                                                                                                                                                                        |
| `P1010`   | "User `${database_user}` was denied access on the database `${database_name}`"<br />                                                                                                                                                                                                                                                                                                                                                                                    |
| `P1014`   | "The underlying \${kind} for model `${model}` does not exist."<br />                                                                                                                                                                                                                                                                                                                                                                                                    |

#### Introspection Engine

| **Error** | **Example message**                                                                      | **Description** |
| :-------- | :--------------------------------------------------------------------------------------- | :-------------- |
| `P4000`   | "Introspection operation failed to produce a schema file: \${introspection_error}"<br /> | `P4001`         | "The introspected database was empty: \${connection_string}"<br /> |
| `P4002`   | "The schema of the introspected database was inconsistent: \${explanation}"<br />        |

### Examples

## <inlinecode>PrismaClientUnknownRequestError</inlinecode>

While the query engine knows many of the possible errors, that can occur, it doesn't know all of them. This error can be an edge-case, that doesn't occur often, which didn't get a code assigned to it yet. While this is an unknown error, this kind of error does not crash the query engine and it can continue running.

### Examples

## <inlinecode>PrismaClientRustPanicError</inlinecode>

In rare cases, the query engine, which is written in Rust, panics. This means, that the engine crashes and exits with a non-zero exit code. In this case, the Prisma Client or the whole Node Process has to be restarted.

### Examples

## <inlinecode>PrismaClientInitializationError</inlinecode>

As soon as prisma.connect() is called or the first request is performed, the Client starts the query engine binary. During this startup of the binary, things can go wrong. Examples:

- The port, on which the query engine http server wants to bind on, can e.g. already be taken.
- The schema might not validate, because an environment variable is missing.

### Examples

## <inlinecode>PrismaClientValidationError</inlinecode>

If the input to a query is incorrect, a Validation Error is being thrown. An example could be a missing field or an incorrect type of a field that is being provided to the Prisma Client.

### Examples
